### PR TITLE
Fix compilation with C++20 

### PR DESCRIPTION
**Pull request by arves100**
Fix compilation for CLI11 to version 2.1.2, required for C++20 support on MSVC.

Changes reference:
https://github.com/CLIUtils/CLI11/pull/136/commits/c93884f1bccc23ae6f4d9ce99723868913012a59

### CODE REVIEW

File: Source/Urho3D/Core/CommandLine.h [hunk 1]
```diff
@@ -55,7 +55,7 @@ namespace CLI
 namespace detail
 {
 
-template <> constexpr const char *type_name<ea::string>() { return "TEXT"; }
+template <> constexpr std::string type_name<ea::string>() { return "TEXT"; }
```
Comment (rokups) [inline Source/Urho3D/Core/CommandLine.h:58]: This change breaks the build. Why did you change it?

Comment (arves100) [reply Source/Urho3D/Core/CommandLine.h:58]: Apperently, MSVC uses the std::string overload and not the "const char*" one,(which is related to the update of CLI11) causing build issues with VS2022.
I haven't tested compilation with other C++ version (that's why I missed the compilation issue), I will check a propriate solution that also works for C++ < 20.

**Comment by michaelaeriksen** [comment]
Any update on this? I would like to use rbfx in a C++20 project.

**Comment by arves100** [comment]
@rokups Sorry for very late 2 year old ping 😅  I've updated the commits which addresses the issues for clang and gcc as well, I've also fixed other small errors that may happen with C++20. It currently buils fine on CLang, GCC and MSVC.
I've also rebased the commit to last version of rbfx.

In regards of embree, I've pushed the commit suggested by https://github.com/embree/embree/pull/400, an update to the last version of embree (which is currenty not released) would properly with C++20 as well, while for tracy I haven't checked if any update would fix as well.

File: Source/Editor/Foundation/ProfilerTab.cpp [hunk 2]
```diff
@@ -119,7 +119,7 @@ void ProfilerTab::RenderContent()
         ui::SameLine();
         bool connect = ui::InputText("", &connectTo_, ImGuiInputTextFlags_EnterReturnsTrue);
         ui::SameLine();
-        connect |= ui::Button(ICON_FA_WIFI " Connect");
+        connect |= ui::Button((const char*)(ICON_FA_WIFI " Connect"));
```
Comment (rokups) [inline Source/Editor/Foundation/ProfilerTab.cpp:122]: Whats the cast for? With or without cast resulting type should still be `const char*`.

File: Source/ThirdParty/embree/common/sys/vector.h [hunk 3]
```diff
@@ -195,7 +195,7 @@ namespace embree
         items = alloc.allocate(new_alloced);
         for (size_t i=0; i<size_active; i++) {
           ::new (&items[i]) T(std::move(old_items[i]));
-          alloc.destroy(&old_items[i]);
+          old_items[i].~T();
```
Comment (rokups) [inline Source/ThirdParty/embree/common/sys/vector.h:198]: This happens to be correct, because all allocators used do exactly this, but i find it weird this is needed. @eugeneko?

File: Source/Urho3D/Core/CommandLine.h [hunk 4]
```diff
@@ -55,7 +55,7 @@ namespace CLI
 namespace detail
 {
 
-template <> constexpr const char *type_name<ea::string>() { return "TEXT"; }
+template <> inline std::string type_name<ea::string>() { return "TEXT"; }
```
[the remark this answers is not in the document]
Comment (arves100) [reply Source/Urho3D/Core/CommandLine.h:58]: I think this was throwing the CI issues last time

File: Source/ThirdParty/embree/common/sys/vector.h (hunk 3 above)
Comment (arves100) [reply Source/ThirdParty/embree/common/sys/vector.h:198]: Embree should push a new version that contains the new changes for C++20, but they don't seem to have pushed it yet so I adopted that patch

**Comment by arves100** [comment]
oh sorry I didn't knew about the main branch repo, I'll be sure to do work in another branch next time xD.
Thank you for reviewing the commits! I have updated the file as you requested and removed the casts, I've tested C++20 compilation with MSVC and it works fine. Let me know if there's anything else I should do

**Review by rokups — APPROVED** [review]

File: Source/Urho3D/Core/CommandLine.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (rokups) [reply Source/Urho3D/Core/CommandLine.h:58]: This one is weird. Looks like CLI11 has such template specializations with both return types. It does indeed not compile with `const char*`.. oh well, not our job to sort out a mess of other project 😁

**Comment by eugeneko** [comment]
Command line parameters don't work for me anymore.
Only simple flags work, the rest of commands crash the application.

**Comment by arves100** [comment]
> Command line parameters don't work for me anymore. Only simple flags work, the rest of commands crash the application.

Ah I see, I'll look at this tomorrow

**Comment by eugeneko** [comment]
Also, I am not sure if (and how) it's related, but Editor UI icons are corrupted
image

**Comment by gleblebedev** [comment]
@eugeneko shall we roll back the change while it gets fixed?

**Comment by eugeneko** [comment]
If we find a fix within a days, probably no. If not -- sure.

**Comment by gleblebedev** [comment]
> If we find a fix within a few days, probably no. If not -- sure.

In a meanwhile shall I keep it rolled back locally? Because I can't run game if --pp doesn't work.
